ndb/src/google/cloud/ndb/model.py

```diff
@@ -697,6 +701,236 @@ def __pos__(self):
+
+        whereas the full list of methods (in order) called here stops once
+        a ``_to_base_type`` method is encountered:
+
+        * ``C._validate()``
+        * ``B._validate()``
+
+        Args:
+            value (Any): The value to be converted / validated.
+
+        Returns:
+            Any: The transformed ``value``.
+        """
+        methods = []
+        for method in self._find_methods("_validate", "_to_base_type"):
+            # Stop if ``_to_base_type`` is encountered.
+            if method.__name__ != "_validate":
+                break
+            methods.append(method)
+
+        call = self._apply_list(methods)
+        return call(value)
+
+    @classmethod
+    def _find_methods(cls, *names, reverse=False):
+        """Compute a list of composable methods.
+
+        Because this is a common operation and the class hierarchy is
+        static, the outcome is cached (assuming that for a particular list
+        of names the reversed flag is either always on, or always off).
+
+        Args:
+            names (Tuple[str, ...]): One or more method names to look up on
+                the current class or base classes.
+            reverse (bool): Optional flag, default False; if True, the list is
+              reversed.
+
+        Returns:
+            List[Callable]: Class method objects.
+        """
+        # Get cache on current class / set cache if it doesn't exist.
+        cache = cls.__dict__.get("_find_methods_cache")
+        if cache:
+            hit = cache.get(names)
+            if hit is not None:
+                return hit
+        else:
+            cache = {}
+            cls._find_methods_cache = cache
+
+        methods = []
+        for klass in cls.__mro__:
+            for name in names:
+                method = klass.__dict__.get(name)
+                if method is not None:
+                    methods.append(method)
+
+        if reverse:
+            methods.reverse()
+
+        cache[names] = methods
+        return methods
+
+    def _apply_list(self, methods):
+        """Chain together a list of callables for transforming a value.
+
+        .. note::
+
+            Each callable in ``methods`` is an unbound instance method, e.g.
+            accessed via ``Property.foo`` rather than ``instance.foo``.
+            Therefore, calling these methods will require ``self`` as the
+            first argument.
+
+        If one of the method returns :data:`None`, the previous value is kept;
+        otherwise the last value is replace.
+
+        Exceptions thrown by a method in ``methods`` are not caught, so it
+        is up to the caller to catch them.
+
+        Args:
+            methods (Iterable[Callable[[Any], Any]]): An iterable of methods
+                to apply to a value.
+
+        Returns:
+            Callable[[Any], Any]: A callable that takes a single value and
+            applies each method in ``methods`` to it.
+        """
+
+        def call(value):
```

Comment:
I am following the original implementation here, but would prefer a class (with `__call__`) encapsulating what is going on here. Though here I realize not all Pythonistas agree with me.

Reply:
If the class would be clearer, or more testable, then go for it (I assume the tests for `_apply_list` would pass either way.

Reply:
I don't think it'd be clearer, so I'll just leave it be (for now). (My biggest testing peeve is that a closure doesn't have a good reference for what scope is enclosed, whereas a `__call__`-able instance does.)

Reply:
Seems like the "closuer doesn't have a good reference" bit affects testability, too.